scipy/interpolate/_regrid_python.py

```diff
@@ -0,0 +1,1110 @@
+     - If growth ended with both axes still minimal: return `return_NdBSpline(...)`.
+     - **Finite-p smoothing (if needed):**
+       - Build 1-D penalty operators: `(Drx, _, ncx) = disc(tx, kx)`,
+         `(Dry, _, ncy) = disc(ty, ky)`, wrap as `PackedMatrix`.
+       - Rebuild design matrices on the final `(tx, ty)`.
+       - Run `_p_search_hit_s(...)` to find `p*` such that `fp(p*) ~ s`:
+         - Internally constructs `F(...)` which evaluates `fp(p)`
+           by calling `_solve_2d_fitpack(...)`.
+         - Uses `root_rati` on `g(p) = fp(p) - s` with the interpolatory
+           reference at `p = inf` and `fp0` bracket.
+       - Return `return_NdBSpline(...)`.
+
+**Separable solver (used by both interpolation and p-search)**
+3. `_solve_2d_fitpack`
+   - Forms augmented banded systems via `_stack_augmented_fitpack`:
+     - X-side: `Ax_aug = [Ax; Dx/p]` if `p != -1`, else just `Ax`.
+     - Y-side: `Ay_aug = [Ay; Dy/p]` if `p != -1`, else just `Ay`.
+     - Pads RHS `Q` with zeros when penalties are stacked.
+   - **Stage X**: `_dierckx.qr_reduce(Ax_aug, ...)` then
+     `_dierckx.fpback(...)` -> intermediate `T`.
+   - Transpose `T` to feed Y solves column-wise; pad if needed.
+   - **Stage Y**: `_dierckx.qr_reduce(Ay_aug, ...)` then
+     `_dierckx.fpback(...)` -> coefficients `C`.
+   - Build CSR design matrices once: `_Ax = Ax.tocsr(...)`, `_Ay = Ay.tocsr(...)`.
+   - Evaluate `zhat = _Ax @ C.T @ _Ay.T`, compute `fp = ||Z_fit - zhat||^2`.
+   - Return `C.T` (in `(nx_coef, ny_coef)` layout) and `fp`.
+
+**Helpers**
+- `_apply_bbox_grid(...)` - slices to `(x_fit, y_fit, Z_fit)`.
+- `build_design_matrices(...)` - wraps `_dierckx.data_matrix` and
+  returns `PackedMatrix` wrappers + `Q`.
+- `_initialise_knots(...)`, `_add_knots(...)` - FITPACK-style knot bookkeeping/growth.
+- `disc(...)` - 1-D roughness operators (packed band form).
+- `F` + `root_rati` - maps `p -> fp(p)` and finds `p*` with `fp(p*) ~ s`.
+- `return_NdBSpline(...)` - packs `(tx, ty, C)` into an `NdBSpline`.
+"""
+import numpy as np
+from scipy.interpolate._ndbspline import NdBSpline
+from scipy.interpolate._fitpack_repro import (
+    root_rati, disc, add_knot, _not_a_knot)
+from . import _dierckx
+from scipy.sparse import csr_array
+
+def ndbspline_call_like_bivariate(ndbs, x, y, dx=0, dy=0, grid=True):
+    """
+    Evaluate a 2D ``NdBSpline`` like a classical bivariate API.
+
+    Parameters
+    ----------
+    ndbs : NdBSpline
+        A 2D spline object (``len(ndbs.t) == 2``).
+    x, y : array_like
+        Sample locations. If ``grid=True``, these must be 1-D strictly
+        increasing vectors. If ``grid=False``, they can be broadcastable
+        arrays of the same shape.
+    dx, dy : int, optional
+        Derivative orders along x and y respectively, by default 0.
+    grid : bool, optional
+        If True, evaluate on the cartesian product of ``x`` and ``y``;
+        otherwise treat ``(x, y)`` as paired coordinates, by default True.
```

Comment:
```suggestion
        If True, evaluate on the cartesian product of `x` and `y`;
        otherwise treat ``(x, y)`` as paired coordinates, by default True.
```